### PR TITLE
Implement integer root

### PR DESCRIPTION
**Pull request by liangjs**

### CODE REVIEW

**Comment by liangjs** [comment]
I am not sure whether `integer_root` or `iroot` is a better name. What do you think?

**Comment by umutoztunc** [comment]
> I am not sure whether `integer_root` or `iroot` is a better name. What do you think?

I am not sure `integer_root` is a nice expressive name. On the other hand, it might be too verbose and one might argue that `iroot` is a common abbreviation since lots of libraries use names like `isqrt` and we even have a function named `ipow` already 😄

File: number_theory/numeric.h [hunk 1]
```diff
@@ -92,10 +95,59 @@ auto pow(T base, U exponent) -> decltype(std::pow(base, exponent)) {
...
+template <typename T,
+          typename U,
+          std::enable_if_t<std::numeric_limits<T>::digits <= 64, bool> = true>
+T iroot(const T &x, const U &n) {
+  static_assert(
+      std::numeric_limits<T>::is_integer && std::numeric_limits<U>::is_integer,
+      "iroot arguments must be integers.");
+  if (n < 0)
+    throw std::invalid_argument("iroot argument |n| should be positive.");
+  if (n == 0)
```
Comment (umutoztunc) [inline number_theory/numeric.h]: What about combining this condition with the one above, making it `if (n <= 0)`? The comments already say that `n` must be positive. Thus, I think `n <= 0` is invalid.

File: number_theory/numeric.h [hunk 2]
```diff
@@ -92,10 +95,59 @@ auto pow(T base, U exponent) -> decltype(std::pow(base, exponent)) {
...
+
+  // For each n, the table stores the maximum y that y^n does not overflow.
+  static const int max_size = 64;
+  static const uint64_t y_limits[max_size] = {
+      0,  0,  4294967295, 2642245, 65535, 7131, 1625, 565, 255, 138, 84, 56, 40,
+      30, 23, 19,         15,      13,    11,   10,   9,   8,   7,   6,  6,  5,
+      5,  5,  4,          4,       4,     4,    3,    3,   3,   3,   3,  3,  3,
+      3,  3,  2,          2,       2,     2,    2,    2,   2,   2,   2,  2,  2,
+      2,  2,  2,          2,       2,     2,    2,    2,   2,   2,   2,  2};
+  uint64_t y_limit = (n >= max_size ? 1 : y_limits[n]);
```
Comment (umutoztunc) [inline number_theory/numeric.h]: I am not sure if the surrounding parentheses contributes to readability here. It might even look cleaner without them. What do you think?

File: number_theory/numeric.h [hunk 3]
```diff
@@ -92,10 +95,59 @@ auto pow(T base, U exponent) -> decltype(std::pow(base, exponent)) {
...
+
+  // Binary search for the maximum y that y^n <= abs(x).
+  uint64_t abs_x = unsigned_abs(x);
```
Comment (umutoztunc) [inline number_theory/numeric.h]: We have an implicit cast here from `make_unsigned_t<T>` to `uint64_t`. I guess we should make it explicit and use `static_cast<uint64_t>` here.

File: number_theory/numeric.h [hunk 4]
```diff
@@ -92,10 +95,59 @@ auto pow(T base, U exponent) -> decltype(std::pow(base, exponent)) {
...
+  uint64_t y_min = 0;
```
Comment (umutoztunc) [inline number_theory/numeric.h:133]: This binary search implementation is not so common. I suggest implementing in the most common way such as:
```cpp
uint64_t left = 0;
uint64_t right = std::min(abs_x, y_limit);
while (left <= right) {
  uint64_t mid = std::mid_point(left, right);
  if (pow(mid, n) <= abs_x)
    left = mid + 1;
  else
    right = mid - 1;
}
return sign(x) * static_cast<T>(left - 1);
```
Usually, people even introduce another variable and update it in the conditional branch which is taken when a valid answer is found which is the case where I updated left above. That way, people avoid returning `left - 1` or `right + 1` in the end. However, I was worried about cost of copying there. Maybe, the cost is negligible.

Another thing I just remembered is that Google's C++ style guide says that if a conditional statement has `else` branch, it should always use curly braces. I don't know if we agree with them or not. I like omitting curlies when the bodies are simple even with multiple cases. Still, it might be good to reconsider our decisions while we are working on this.

File: number_theory/numeric_unittest.cpp [hunk 5]
```diff
@@ -81,5 +81,39 @@ TEST(NumericTest, Pow) {
...
+  // test small numbers
+  EXPECT_EQ(iroot(T(120), 2), 10);
+  EXPECT_EQ(iroot(T(125), 3), 5);
+  EXPECT_EQ(iroot(T(0), 1), 0);
+  if (std::numeric_limits<T>::is_signed) {
+    EXPECT_EQ(iroot(T(-125), 3), -5);
+    EXPECT_EQ(iroot(T(-30), 3), -3);
+  }
+
+  // test exceptions
```
Comment (umutoztunc) [inline number_theory/numeric_unittest.cpp:95]: I think we should also test `std::invalid_argument` exception with a negative `n` case to cover all the execution paths in the tests.

**Review by umutoztunc — CHANGES_REQUESTED** [review]
Nice work! I just did some nitpicking. I am looking forward to merging this!

File: number_theory/numeric_unittest.cpp (hunk 5 above)
Comment (liangjs) [reply number_theory/numeric_unittest.cpp:95]: I think we don't need to test undocumented exceptions? Like we didn't test invalid_argument in inverse_mod.

Comment (umutoztunc) [reply number_theory/numeric_unittest.cpp:95]: I think we should test it. I did not realize that I missed it during tests. The point is to test all execution paths which is called whitebox testing. The reason we do not document the exception is that it contradicts with the documented conditions. Thus, documenting a contradicting thing makes it look like it is part of the API. It is like we are suggesting them that it should be used that way. In reality, for example, the mod should never be negative. It is not a feature that we throw exception when they provide negative modulus. However, we still need to test it to make sure that it does not break.

File: number_theory/numeric.h (hunk 4 above)
Comment (liangjs) [reply number_theory/numeric.h:133]: There are many ways to implement binary search, and I don't think one way is more common than another. Actually the version you provided does not look better, because it has many `+1` and `-1` which confuse me.

Comment (liangjs) [reply number_theory/numeric.h:133]: I am looking forward to switch to `std::ranges::binary_search` once this c++20 feature is supported by clang-12.

Comment (umutoztunc) [reply number_theory/numeric.h:133]: > There are many ways to implement binary search, and I don't think one way is more common than another. Actually the version you provided does not look better, because it has many `+1` and `-1` which confuse me.

It is interesting that you are not used to this version. I thought that was the most common one since that's what I saw in almost any tutorial and book. I was using your version during codeforces competitions but then I realized most editorials had the version I posted above, including red coders or even LGMs. I am not sure about it either, but even Errichto seems to prefer that version:
https://www.youtube.com/watch?v=GU7DpgHINWQ

> I am looking forward to switch to `std::ranges::binary_search` once this c++20 feature is supported by clang-12.

This sounds good, but I would also consider `std::ranges::lower_bound` since we are not searching for a specific value.

Comment (liangjs) [reply number_theory/numeric.h:133]: I submitted a commit for a new style, although it looked more confusing to me. You can decide whether it is better and which way we go for.

**Comment by liangjs** [comment]
I deleted the invalid_argument tests and switched back to my binary search style, as we discussed.

**Review by umutoztunc — APPROVED** [review]
